src/clients/HttpTunneler.cc

```diff
@@ -329,13 +342,48 @@ Http::Tunneler::bailWith(ErrorState *error)
 {
     Must(error);
     answer().squidError = error;
+
+    if (CachePeer *p = connection->getPeer())
+        peerConnectFailed(p);
+
+    callBack();
+    disconnect(true);
+}
+
+void
+Http::Tunneler::sendSuccess()
+{
+    assert(answer().positive());
     callBack();
+    disconnect(false);
+}
+
+void
+Http::Tunneler::disconnect(const bool andClose)
+{
+    if (!connection)
+        return;
+
+    // remove close handler
+    comm_remove_close_handler(connection->fd, closer);
+    closer = nullptr;
+
+    // remove connection timeout handler
+    commUnsetConnTimeout(connection);
+
+    if (andClose) {
+        if (usesPconn_)
+            fwdPconnPool->noteUses(fd_table[connection->fd].pconn.uses);
+        connection->close();
+        connection = nullptr;
+    }
 }
 
 void
 Http::Tunneler::callBack()
 {
     debugs(83, 5, connection << status());
+    answer().conn = connection;
```

Comment:
Do we really have to set `.conn` on negative answers? Would it be possible to restrict this answer field to good/usable connections?

Reply:
Yes we can.
However on negative answers the connection is nil so the result is the same.

commit 748a0a5cf2